File: Part1_강의자료/Part1_쪽지시험/문제별코드/ProblemD_회전초밥/15961.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

int main()
{
    int N, D, K, C;
    scanf("%d %d %d %d", &N, &D, &K, &C);

    vector<int> vec(N);
    for (int i = 0; i < N; i++)
        scanf("%d", &vec[i]);

    vector<int> dish_count(D + 1);
    int kind = 0;
    for (int i = 0; i < K - 1; i++)
        if (dish_count[dist[i]]++ == 0)
            kind++;

    int ans = 0;
    for (int i = 0; i < N; i++)
    {
        if (dish_count[dish[(i + K - 1) % N]]++ == 0)
            kind++;
        ans = max(ans, kind + (dish_count[C] == 0 ? 1 : 0));
        if (--dish_count[dish[i]] == 0)
            kind--;
    }

    printf("%d\n", ans);
    return 0;
}

File: Part1_강의자료/Part1_쪽지시험/문제별코드/ProblemD_회전초밥/15961.py
import sys
input = sys.stdin.readline

N, D, K, C = map(int, input().split())
dish = [int(input()) for _ in range(N)]
dish_count = [0] * (D + 1)
kind = 0

for i in range(K - 1):
    if dish_count[dish[i]] == 0:
        kind += 1
    dish_count[dish[i]] += 1

ans = 0
for i in range(N):
    if dish_count[dish[(i + K - 1) % N]] == 0:
        kind += 1
    dish_count[dish[(i + K - 1) % N]] += 1
    ans = max(ans, kind + (1 if dish_count[C] == 0 else 0))
    
    dish_count[dish[i]] -= 1
    if dish_count[dish[i]] == 0:
        kind -= 1

print(ans)
